JavaScript development/Frameworks/React.md

```diff
@@ -0,0 +1,536 @@
+    // ...
+  }
+
+  render () {
+    return (
+      <ChildComponent onClick={this.onClickHandler} />
+    );
+  }
+}
+```
+
+Omit values for boolean props if default value is **true**.
+
+``` jsx
+// bad
+<ChildComponent
+  onClick={this.onClickHandler}
+  isPropActive={true}
+/>
+
+// good
+<ChildComponent
+  onClick={this.onClickHandler}
+  isPropActive
+/>
+```
+
+### Receiving Props
+*Checkout [`prop-types`](https://github.com/facebook/prop-types) lib on GitHub for general info.*
+
+In class components:
+- always use prop validation
+- use **static member** for declaring prop types and default props (if your transpiler doesn't support static members, set them after class declaration and then export the class; See: ["Exporting Components" section](#exporting-components))
+- use *defaultProps* if you need to set default values
+
+``` jsx
+export default class ComponentName extends React.Component {
+  static propTypes = {
+    id: PropTypes.string.isRequired,
+    name: PropTypes.string,
+  };
+
+  static defaultProps = {
+    name: 'John Doe',
+  };
+
+  // ...
+
+  render() {
+    const {name} = this.props;
+
+    return (
+      <div>{name}</div>
+    );
+  }
+}
+```
+
+In stateless components, use PropTypes for type checking but use native JS for setting default values.
+
+``` jsx
+function ComponentName({id, name = 'John Doe'}) {
+  // ...
+};
+
+ComponentName.propTypes = {
+  id: PropTypes.number.isRequired,
+  name: PropTypes.string
+}
+```
+
+### Using props
+
+If you are using more than one property in a method, preffer accessing them with **object destructuring**.
+
+``` jsx
+class ComponentName extends React.Component {
+  // ...
+  render() {
+    const {id, name, onClickHandler} = this.props;
+    // ...
+  }
+}
+```
+
+### Importing React
+``` jsx
+import * as React from 'react';
+
+class ComponentName extends React.Component {
+  // ...
+}
+```
+
+### Exporting Components
+Each React component file should have a single exported component.
+ 
+Exporting class components:
+``` jsx
+// preferred way of exporting
+export default class ComponentName extends React.Component {
+  // ...
+}
+
+// exporting when static members are not supported by the transpiler:
+class ComponentName extends React.Component {
+  // ...
+}
+
+ComponentName.propTypes = {
+  // ..
+};
+
+export default ComponentName;
+```
+
+Exporting stateless components:
+``` jsx
+// stateless components
+function PureComponent(/* props */) {
+  // ...
+}
+
+PureComponent.propTypes = {
+  // ...
+};
+
+export default PureComponent;
+
+// stateless components with no prop validation
+export default () => (/* ... */);
+```
+
+### Rendering Arrays of Data
+
+For iterating arrays:
+- use **map** method to render items
+- you should have a unique key for each element
+- **don't use array's index as a key**
+
+``` jsx
+render() {
+  // ...
+  return (
+    <ul>
+      {
+        booksArray.map((book) => {
+          <ListItem key={book.id} title={book.title} /> 
```

Comment:
There should also be a different example - an example where the items have nested elements, e.g. a link inside of a `<li>`